src/network_event/client_event.rs

```diff
@@ -119,155 +119,219 @@ impl ClientEventAppExt for App {
+        let events = world.get_resource::<Events<T>>().unwrap();
+
+        for event in events.get_reader().read(&events) {
+            let message = DefaultOptions::new()
+                .serialize(&event)
+                .expect("mapped client event should be serializable");
+
+            trace!("Sending event: {}", std::any::type_name::<T>());
+            client.send(channel_id, message)
+        }
+    })
+}
+
 /// Transforms `T` events into [`FromClient<T>`] events to "emulate"
 /// message sending for offline mode or when server is also a player.
-fn resend_locally<T: Event>(
-    mut events: ResMut<Events<T>>,
-    mut client_events: EventWriter<FromClient<T>>,
-) {
-    for event in events.drain() {
-        client_events.send(FromClient {
-            client_id: ClientId::SERVER,
-            event,
-        });
-    }
+fn resend_locally<T: Event + Serialize>(world: &mut World) {
+    world.resource_scope(|world, mut events: Mut<Events<T>>| {
+        world.resource_scope(|_world, mut client_events: Mut<Events<FromClient<T>>>| {
+            if events.len() > 0 {
+                let mapped_events = events.drain().map(|event| FromClient {
+                    client_id: ClientId::SERVER,
+                    event,
+                });
+                trace!("Resending event: {}", std::any::type_name::<T>());
+
+                client_events.send_batch(mapped_events);
+            }
+        })
+    })
+}
+
+fn receive<T: Event + DeserializeOwned>(world: &mut World, channel_id: u8) {
+    world.resource_scope(|world, mut server: Mut<RepliconServer>| {
+        world.resource_scope(|_world, mut client_events: Mut<Events<FromClient<T>>>| {
+            for (client_id, message) in server.receive(channel_id) {
+                match DefaultOptions::new().deserialize(&message) {
+                    Ok(event) => {
+                        trace!(
+                            "applying event `{}` from `{client_id:?}`",
+                            std::any::type_name::<T>()
+                        );
+                        client_events.send(FromClient { client_id, event });
+                    }
+                    Err(e) => debug!("unable to deserialize event from {client_id:?}: {e}"),
+                }
+            }
+        })
+    })
 }
 
 /// Discards all pending events.
 ///
 /// We discard events while waiting to connect to ensure clean reconnects.
-fn reset<T: Event>(mut events: ResMut<Events<T>>) {
-    let drained_count = events.drain().count();
-    if drained_count > 0 {
-        warn!("discarded {drained_count} client events due to a disconnect");
+fn reset<T: Event>(world: &mut World) {
+    world.resource_scope(|_world, mut events: Mut<Events<T>>| {
+        let drained_count = events.drain().count();
+        if drained_count > 0 {
+            warn!("Discarded {drained_count} client events due to a disconnect");
+        }
+    })
+}
+
+#[derive(Resource, Default)]
+struct ClientEventRegistry {
```

Comment:
Same with the registry: I would move it upper and store both type of events in a single resource, you won't need parallelization here.